### PR TITLE
[FEATURE] Chatroom offline character persistence

### PR DESCRIPTION
**Pull request by Jomshir98**
Works towards: #70 
Based on following proposal: [Pandora Discord channel link](https://discord.com/channels/872284471611760720/1025099479877169233/1121746790849449985)

This PR implements part of the proposal, to be specific it does the following:
- Characters now always stay in rooms unless they leave manually
- Characters and rooms get properly loaded and unloaded as needed, if room is loaded, it always loads all characters that are inside
- Fixed bug with directory could crash when stopping with connected shards
- Updated displaying online characters (both in contacts and in own character list) to only show when the character is actually selected by player
- While at it fixed bug that new friends show as offline in the list until refresh
- Show "Offline" status next to offline characters in the "Room" tab character list
- Show offline characters differently in graphics (there is a new "Interface" option to change what effect is used, defaults to "Ghost" (darken + semi-transparent)
- Added a warning to the leave dialog, if inside a room, that character will stay in it even on character change/logout
- Fixed characters being removed from room devices when room unloads or is moved to another shard

Known bugs:
- ~~Characters are still randomly removed from room devices when the room gets unloaded~~ _fixed_

Related things NOT implemented as part of this, will be done in later PRs (unless we decide to delay this until said feature is done):
- Timer to auto-kick/auto-leave the room when offline - characters are only removed based on manual action
- Chat messages storing logic for offline characters (currently offline characters do not store messages at all)
- Any extra chat messages around disconnecting/reconnected characters

### CODE REVIEW

File: pandora-server-directory/src/database/mongoDb.ts [hunk 1]
```diff
@@ -325,7 +359,7 @@ export default class MongoDatabase implements PandoraDatabase {
 		return await this._chatrooms.find({
 			owners: { $elemMatch: { $in: [account] } },
 		})
-			.project<Pick<IChatRoomDirectoryData, 'id' | 'config' | 'owners'>>({ id: 1, config: 1, owners: 1 })
+			.project<Pick<IChatRoomDirectoryData, 'id' | 'config' | 'owners' | 'accessId'>>({ id: 1, config: 1, owners: 1, accessId: 1 })
```
Comment (Sekkmer) [inline pandora-server-directory/src/database/mongoDb.ts:362]: can we use `typeof CHATROOM_DIRECTORY_PROPERTIES[number]` here?

File: pandora-client-web/src/components/chatroom/chatRoomCharacter.tsx [hunk 2]
```diff
@@ -124,23 +127,35 @@ function ChatRoomCharacterDisplay({
...
+		position: dataPosition,
+		settings,
+		isOnline,
+	} = useCharacterData(character);
+
+	const { interfaceChatroomOfflineCharacterFilter } = useCurrentAccountSettings();
+
+	const playerFilters = usePlayerVisionFilters(character.isPlayer());
+	const characterFilters = useCharacterDisplayFilters(character);
+	const filters = [...playerFilters, ...characterFilters];
```
Comment (Sekkmer) [inline pandora-client-web/src/components/chatroom/chatRoomCharacter.tsx:142]: shouldn't this be a `useMemo`?

**Review by Sekkmer — CHANGES_REQUESTED** [review]

File: pandora-server-directory/src/account/accountRelationship.ts [hunk 3]
```diff
@@ -85,9 +87,11 @@ export class AccountRelationship {
 		}
 		if (this.account.data.settings.allowDirectMessagesFrom === 'room') {
 			for (const char of this.account.characters.values()) {
-				if (!char.room) continue;
+				const room = char?.loadedCharacter?.room;
```
[the remark this answers is not in the document]
Comment (Jomshir98) [reply pandora-server-directory/src/account/accountRelationship.ts:88]: I'm not quite sure what you mean? To me it is better keeping only one source of truth, even if it means checking slightly more data than needed and in this case it feels minimal - from the old behaviour it changed only by adding the middle step of `loadedCharacter`. Also I'm not sure if the `char` can actually ever be null or if the nullcheck is there only to stop TS from complaining.

File: pandora-server-directory/src/account/character.ts [hunk 4]
```diff
@@ -155,208 +99,396 @@ export class Character {
...
+		return Promise.resolve();
+	}
+}
+
+export type CharacterAssignment = {
+	// Character is assigned to a specific shard and is not in any room
+	type: 'shard';
+	shard: Shard;
+} | {
+	// Character is assigned to the same shard the room is, but is not in the room
```
[the remark this answers is not in the document]
Comment (Jomshir98) [reply pandora-server-directory/src/account/character.ts:182]: This is described a bit obscurely in room join:
> Must connect to the same shard as the room to check character-based join requirements
I will add comment that this is necessary for the flow of joining a room.

File: pandora-server-directory/src/database/mongoDb.ts (hunk 1 above)
Comment (Jomshir98) [reply pandora-server-directory/src/database/mongoDb.ts:362]: I would agree with doing that if the pick filter object would also be autogenerated. I stuck with doing it manually as then TS catches that (for mongo the object can be any record, it doesn't check the exact property names, making it unsafe without being this explicit)

File: pandora-client-web/src/components/chatroom/chatRoomDevice.tsx [hunk 5]
```diff
@@ -375,11 +375,13 @@ function RoomDeviceGraphicsLayerSlot({ item, layer }: {
...
 	layer: IRoomDeviceGraphicsLayerSlot;
-	character: Character;
+	character: Character<ICharacterRoomData>;
 	characterState: AssetFrameworkCharacterState;
 }): ReactElement | null {
 	const debugConfig = useDebugConfig();
-	const filters = usePlayerVisionFilters(character.isPlayer());
+	const playerFilters = usePlayerVisionFilters(character.isPlayer());
+	const characterFilters = useCharacterDisplayFilters(character);
+	const filters = [...playerFilters, ...characterFilters];
```
Comment (Sekkmer) [inline pandora-client-web/src/components/chatroom/chatRoomDevice.tsx:384]: here too, I just forgot to mark it last time

**Review by Sekkmer — APPROVED** [review]

**Pull request merged**